### PR TITLE
BDD Test Suite

### PR DESCRIPTION
**Pull request by CharlesRendle**
A Behave style test suite for databaker functionality.

Includes tests for all bag overrides except: group, parent and children
And tests the whole "databaking" process (taking a dataset from file to transformed dataframe)

I have had to add 1 line to `overrides.py` because of an issue with python 3.7 and `re._pattern_type`. See github issue ticket for further details.

### CODE REVIEW

**Review by mikeAdamss — COMMENTED** [review]
first, great job in getting this all in and working! :muscle: 

I'm not gonna comment line by line, but there are some behave/python tricks that can massively simplify this for you and reduce the amount of code significantly (also removes my nasty temporary python hack from months ago).

as an example, I've added a branch with a refactored test for `Complete the databaking process - produce a correct dataframe`
https://github.com/GSS-Cogs/databaker/compare/bdd-tests...GSS-Cogs:bdd-tests-refactor-example?expand=1 

if I'm understanding correctly, we can capture most of what we need with just those steps + passing variables in.

there's a lot going on here, so grab me if I'm off base or you want a chat through it.

File: features/temp_transform.py [hunk 1]
```diff
@@ -0,0 +1,6 @@
+from databaker.framework import *
```
Comment (mikeAdamss) [inline features/temp_transform.py]: this whole file is a leftover from the previous approach, you can remove it.

**Review by mikeAdamss — COMMENTED** [review]
Looking really good now, nice 😄 

This is my inner pedant coming out but as well as getting rid of that file (see below), can we remove the additional whitespace lines in the scenarios within the feature files please, it should be a blank line _between_ each scenario not each step. Not sure the reasons but that's how they're formatted usually and how they're formatted in our other repos.

With those tweaks in am happy to approve and merge 👍

File: features/fixtures/bakingtesttransform.py [hunk 2]
```diff
@@ -0,0 +1,70 @@
+# ---
```
Comment (mikeAdamss) [inline features/fixtures/bakingtesttransform.py]: not sure what this file is, do we still need it?

File: features/fixtures/.ipynb_checkpoints/bakingtesttransform-checkpoint.py [hunk 3]
```diff
@@ -0,0 +1,70 @@
+# ---
```
Comment (mikeAdamss) [inline features/fixtures/.ipynb_checkpoints/bakingtesttransform-checkpoint.py]: not sure what this file is, do we still need it?

File: features/temp_transform.py (hunk 1 above)
Comment (CharlesRendle) [reply features/temp_transform.py]: Removed

File: features/fixtures/bakingtesttransform.py (hunk 2 above)
Comment (CharlesRendle) [reply features/fixtures/bakingtesttransform.py]: This was to test the homemade dataset which displays all the direction qualities we wanted to test. We don't need it as the transform code is brought into baking.feature
Removed

File: features/fixtures/.ipynb_checkpoints/bakingtesttransform-checkpoint.py (hunk 3 above)
Comment (CharlesRendle) [reply features/fixtures/.ipynb_checkpoints/bakingtesttransform-checkpoint.py]: Leftover from testing the homemade dataset using jupyter lab. We don't need it so has been removed.

File: features/isnotwhitespace.feature [hunk 4]
```diff
@@ -0,0 +1,16 @@
...
+  I want to be able to define a dimension as a range of values from a cell reference, but I want to
+  ignore any whitespace cells that lie within that range.
+
+  Scenario: Define non-whitespace year from a range of cell references ignoring any whitespace cells.
+    Given we load an xls file named "bulletindataset2v2.xls"
+    And select the sheet "Table 2a"
+    And we define cell selections as
+      | key             | value                                             |  
+      | unit            | tab.excel_ref("A11:A250").is_not_whitespace()     |
+    Then we confirm year contains no blank cells.
```
Comment (mikeAdamss) [inline features/isnotwhitespace.feature]: think this is an old approach leftover and reads a bit odd, I'm guessing i needs to be `we confirm "unit" contains no blank cells`.

in the step you can do:

```
@then(u'we confirm "{selection_name}" contains no blank cells.')
def step_impl(context, selection_name):
    bag_of_cells = context.selections[selection_name]
    # check whatever it is you need to check about the "bag_of_cells" selection
```

**Review by mikeAdamss — COMMENTED** [review]
this is looking really good now - but, I spotted an old approach leftover and my brain won't let me press approve 😄

File: features/isnotwhitespace.feature (hunk 4 above)
Comment (CharlesRendle) [reply features/isnotwhitespace.feature]: Great spot, knew I forgot something. 
I have changed the step to refer to `selection name` which is in keeping with the existing steps. The actual name can always be found from the key in the table attached to the definition step. I have changed most of these to reflect this.

**Comment by mikeAdamss** [comment]
looks good, good job 👍 . merge away 😄

**Pull request merged**